http/oidc/src/main/java/org/wildfly/security/http/oidc/OidcRequestAuthenticator.java

```diff
@@ -416,4 +484,112 @@ private static boolean hasScope(String scopeParam, String targetScope) {
         }
         return false;
     }
+    private String convertToRequestParameter(URIBuilder redirectUriBuilder, String redirectUri, String state, List<NameValuePair> forwardedQueryParams) throws Exception {
+        redirectUriBuilder.addParameter(SCOPE, OIDC_SCOPE);
+        forwardedQueryParams.add(new BasicNameValuePair(STATE, state));
+        forwardedQueryParams.add(new BasicNameValuePair(REDIRECT_URI, redirectUri));
+        forwardedQueryParams.add(new BasicNameValuePair(RESPONSE_TYPE, CODE));
+        forwardedQueryParams.add(new BasicNameValuePair(CLIENT_ID, deployment.getResourceName()));
+
+        JwtClaims jwtClaims = new JwtClaims();
+        jwtClaims.setIssuer(deployment.getIssuerUrl());
+        jwtClaims.setAudience(deployment.getProviderUrl());
+        for ( NameValuePair parameter: forwardedQueryParams) {
+            jwtClaims.setClaim(parameter.getName(), parameter.getValue());
+        }
+
+        // sign JWT first before encrypting
+        JsonWebSignature signedRequest = signRequest(jwtClaims);
+
+        // Encrypting optional
+        if (deployment.getRequestEncryptAlgorithm() != null && !deployment.getRequestEncryptAlgorithm().isEmpty() &&
+            deployment.getRequestEncryptEncValue() != null && !deployment.getRequestEncryptEncValue().isEmpty()) {
+            return encryptRequest(signedRequest).getCompactSerialization();
+        } else {
+            return signedRequest.getCompactSerialization();
+        }
+    }
+
+    public KeyPair getkeyPair() throws IOException {
+        if (deployment.getClientKeyStoreFile().contains(PROTOCOL_CLASSPATH)) {
+            deployment.setClientKeyStoreFile(deployment.getClientKeyStoreFile().replace(PROTOCOL_CLASSPATH, Objects.requireNonNull(Thread.currentThread().getContextClassLoader().getResource("")).getPath()));
+        }
+        if (!deployment.getRequestSignatureAlgorithm().equals(NONE) && deployment.getClientKeyStoreFile() == null){
+            throw log.invalidKeyStoreConfiguration();
+        } else {
+            return loadKeyPairFromKeyStore(deployment.getClientKeyStoreFile(),
+                    deployment.getClientKeyStorePassword(), deployment.getClientKeyPassword(),
+                    deployment.getClientKeyAlias(), deployment.getClientKeystoreType());
+        }
+    }
+
+    public JsonWebSignature signRequest(JwtClaims jwtClaims) throws Exception {
+        JsonWebSignature jsonWebSignature = new JsonWebSignature();
+        jsonWebSignature.setPayload(jwtClaims.toJson());
+
+        if (deployment.getRequestSignatureAlgorithm().contains(NONE)) { //unsigned
+            jsonWebSignature.setAlgorithmConstraints(AlgorithmConstraints.NO_CONSTRAINTS);
```

Comment:
I think this is the default value. Do we need to specify it?

Reply:
I think we do, because if I comment it out, I get "Signature algorithm header (alg) not set." exception